src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1013 @@
+  GPR_DEBUG_ASSERT(initialized_);
+  // host_ and port_ should have been parsed successfully in Initialize.
+  const std::string hostport = grpc_core::JoinHostPort(host_, port_);
+  // TODO(yijiem): maybe add ResolvedAddress version of these to
+  // tcp_socket_utils.cc
+  grpc_resolved_address addr;
+  if (grpc_parse_ipv4_hostport(hostport, &addr, false /* log errors */) ||
+      grpc_parse_ipv6_hostport(hostport, &addr, false /* log errors */)) {
+    Result result;
+    result.emplace_back(reinterpret_cast<sockaddr*>(addr.addr), addr.len);
+    event_engine_->Run([on_resolve = std::move(on_resolve_),
+                        result = std::move(result)]() mutable {
+      grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+      grpc_core::ExecCtx exec_ctx;
+      on_resolve(std::move(result));
+    });
+    return true;
+  }
+  return false;
+}
+
+void GrpcAresHostnameRequest::LogResolvedAddressesList(
+    const char* input_output_str) {
+  for (size_t i = 0; i < result_.size(); i++) {
+    auto addr_str = ResolvedAddressToString(result_[i]);
+    gpr_log(GPR_INFO,
+            "(ares driver) request:%p c-ares address sorting: %s[%" PRIuPTR
+            "]=%s",
+            this, input_output_str, i,
+            addr_str.ok() ? addr_str->c_str()
+                          : addr_str.status().ToString().c_str());
+  }
+}
+
+void GrpcAresHostnameRequest::SortResolvedAddresses() {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_trace_ares_wrapper_address_sorting)) {
+    LogResolvedAddressesList("input");
+  }
+  address_sorting_sortable* sortables = static_cast<address_sorting_sortable*>(
+      gpr_zalloc(sizeof(address_sorting_sortable) * result_.size()));
+  for (size_t i = 0; i < result_.size(); i++) {
+    sortables[i].user_data = &result_[i];
+    memcpy(&sortables[i].dest_addr.addr, result_[i].address(),
+           result_[i].size());
+    sortables[i].dest_addr.len = result_[i].size();
+  }
+  address_sorting_rfc_6724_sort(sortables, result_.size());
+  Result sorted;
+  sorted.reserve(result_.size());
+  for (size_t i = 0; i < result_.size(); i++) {
+    sorted.emplace_back(
+        *static_cast<EventEngine::ResolvedAddress*>(sortables[i].user_data));
+  }
+  gpr_free(sortables);
+  result_ = std::move(sorted);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_trace_ares_wrapper_address_sorting)) {
+    LogResolvedAddressesList("output");
+  }
+}
+
+GrpcAresSRVRequest::GrpcAresSRVRequest(
+    absl::string_view name, EventEngine::Duration timeout,
+    std::unique_ptr<GrpcPolledFdFactory> polled_fd_factory,
+    EventEngine* event_engine)
+    : GrpcAresRequest(name, absl::nullopt, timeout,
+                      std::move(polled_fd_factory), event_engine) {}
+
+GrpcAresSRVRequest::~GrpcAresSRVRequest() {
+  GRPC_ARES_WRAPPER_TRACE_LOG("request:%p destructor", this);
+}
+
+void GrpcAresSRVRequest::Start(
+    absl::AnyInvocable<void(absl::StatusOr<Result>)> on_resolve) {
+  auto self = Ref(DEBUG_LOCATION, "Start");
```

Comment:
Yep, makes sense. And the ref needs to be taken here (not below next to `ares_query` where it's actually necessary), since the final unref should not happen while holding the lock. Please leave a comment to that effect so this line does not inadvertently move.